treeherder/etl/taskcluster_pulse/handler.py

```diff
@@ -0,0 +1,295 @@
+# Code imported from https://github.com/taskcluster/taskcluster/blob/32629c562f8d6f5a6b608a3141a8ee2e0984619f/services/treeherder/src/handler.js
+import logging
+
+import jsonschema
+import slugid
+import taskcluster
+import taskcluster_urls
+
+from treeherder.etl.schema import get_json_schema
+from treeherder.etl.taskcluster_pulse.artifact_links import addArtifactUploadedLinks
+from treeherder.etl.taskcluster_pulse.parse_route import parseRoute
+
+logger = logging.getLogger(__name__)
+root_url = "https://taskcluster.net"
+
+# Build a mapping from exchange name to task status
+EXCHANGE_EVENT_MAP = {
+    "exchange/taskcluster-queue/v1/task-pending": "pending",
+    "exchange/taskcluster-queue/v1/task-running": "running",
+    "exchange/taskcluster-queue/v1/task-completed": "completed",
+    "exchange/taskcluster-queue/v1/task-failed": "failed",
+    "exchange/taskcluster-queue/v1/task-exception": "exception",
+}
+
+
+class PulseHandlerError(Exception):
+    """Base error"""
+    pass
+
+
+def stateFromRun(jobRun):
+    return "completed" if jobRun["state"] in ("exception", "failed") else jobRun["state"]
+
+
+def resultFromRun(jobRun):
+    if jobRun["state"] == "completed":
```

Comment:
This seems like it could avoid a lot of these ``elif``s by creating a map for ``state`` and ``reasonResolved``.  You could have a map that contains all of "completed", "success", "fail", "canceled" and "superseded".  Then check if ``reasonResolved in map`` exists.  If not, look for ``state in map``, otherwise return "unknown".

Just a thought.  But if you prefer this, fine by me.  :) 

But it seems like at least extracting ``state`` and ``reasonResolved`` from ``jobRun`` so you're not doing it each time, would be preferable.